src/baseline.c

```diff
@@ -352,6 +353,137 @@ s8 lesq_solution_float(u8 num_dds_u8, const double *dd_obs, const double *N,
   return 0;
 }
 
+static void drop_i(u32 index, u32 len, u32 size, const double *from, double *to)
+{
+  memcpy(to, from, index*size*sizeof(double));
+  memcpy(to + index*size, from + (index + 1)*size, (len - index - 1)*size*sizeof(double));
+}
+
+static s8 lesq_without_i(u8 dropped_dd, u8 num_dds, const double *dd_obs,
+                  const double *N, const double *DE, double b[3], double *resid)
+{
+  u8 new_dds = num_dds - 1;
+  double new_obs[new_dds];
+  double new_N[new_dds];
+  double new_DE[new_dds * 3];
+
+  drop_i(dropped_dd, num_dds, 1, dd_obs, new_obs);
+  drop_i(dropped_dd, num_dds, 1, N, new_N);
+  drop_i(dropped_dd, num_dds, 3, DE, new_DE);
+
+  return lesq_solution_float(new_dds, new_obs, new_N, new_DE, b, resid);
+}
+
+/* Approximate chi square test
+ * Scales least squares residuals by estimated covariance, compares against threshold.
+ * Returns true if norm is below threshold.
+ *
+ * Returns residual value in `residual' if non-null.
+ */
+static bool chi_test(u8 num_dds, double *residuals, double *residual)
+{
+  double sigma = DEFAULT_PHASE_VAR_KF;
+  /* 5.5 seems adequate for float/fixed baseline residuals. */
+#define BASELINE_RESIDUAL_THRESHOLD 5.5
+  double norm = vector_norm(num_dds, residuals) / sqrt(sigma);
+  if (residual) {
+    *residual = norm;
+  }
+  return norm < BASELINE_RESIDUAL_THRESHOLD;
+}
+
+/* See lesq_solution_float for argument documentation
+ * Return values:
+ *    0: solution with all dd's ok
+ *    1: repaired solution, using one fewer observation
+ *       returns index of removed observation if removed_obs ptr is passed
+ *   -1: no reasonable solution possible
+ */
+/* TODO(dsk) update all call sites to use n_used as calculated here.
+ * TODO(dsk) add warn/info logging to call sites when repair occurs. */
+s8 lesq_solve_and_check(u8 num_dds_u8, const double *dd_obs,
+                        const double *N, const double *DE, double b[3],
+                        u8 *n_used,
+                        double *ret_residuals,
+                        u8 *removed_obs)
+{
+  integer num_dds = num_dds_u8;
+  double residuals[num_dds];
+
+  s8 okay = lesq_solution_float(num_dds_u8, dd_obs, N, DE, b, residuals);
+
+  if (okay == 0) {
```

Comment:
I'd invert this if condition so it doesn't have to wrap around the whole function, i.e.

```
if (okay != 0) {
    /* Not enough sats or other error returned by initial lesq solution. */
    return -1;
}
```